kafka-client/src/main/java/dev/responsive/kafka/store/CommitBuffer.java

```diff
@@ -402,13 +405,19 @@ private RemoteWriteResult drain(final Collection<RemoteWriter<K>> writers) {
   }
 
   private void maybeTruncateChangelog(final long offset) {
-    if (truncateChangelog) {
+    if (truncateChangelog && !isNotDeleteEnabled) {
       try {
         admin.deleteRecords(Map.of(changelog, RecordsToDelete.beforeOffset(offset))).all().get();
         log.info("Truncated changelog topic {} before offset {}", changelog, offset);
       } catch (final ExecutionException e) {
-        log.warn("Could not truncate changelog topic-partition {}.", changelog, e);
+        log.warn("Could not truncate changelog topic-partition " + changelog, e);
+        if (e.getCause() instanceof PolicyViolationException) {
+          log.info("Disabling further changelog truncation attempts due to topic configuration "
+                       + "being incompatible with deleteRecord requests", e);
+          isNotDeleteEnabled = true;
```

Comment:
is the reason for doing this in the exception handler to avoid admin calls during store initialization?

Reply:
Yeah, basically. Most cases I think it should be unnecessary to check at all, so doing it on every store init is pretty costly. Ideally we could just check once upfront, when the app first starts up, but atm there's not really a good way/place to do this, because initialization is the first place where we have all the necessary ingredients all together: the changelog name and the store params plus the admin (which is extracted through the configs that only get passed in to #init)

Anyways, I think it's always possible for users to change their topic cleanup.policy on-the-fly, so they could hit this exception even if they passed any startup checks. I think it's better to just permanently disable changelog truncation in a case like that, rather than cripple the application with all these failing requests. If they want to re-enable changelog truncation after changing the topic configs back, they can simply bounce the app. That kind of double-turnaround seems particularly unlikely so I feel like a restart is a warranted solution